File: apps/api/src/verify/verify.service.ts
import { Injectable } from '@nestjs/common';
import { SupabaseService } from '../supabase/supabase.service';
import { StorageService } from '../storage/storage.service';
import { SijilliErrors } from '../common/errors/error-envelope';

// Public verification of a deed by property_code. The response is a
// SANITIZED view: only the citizen's first and family names are
// returned, middle names are masked, and PII like phone/dob never
// leaves the API. The endpoint is unauthenticated by design — verify
// QR codes are public.

export interface PublicDeedView {
  property_code: string;
  parcel_number: string | null;
  property_type: string;
  area_sqm: number | null;
  status: 'active' | 'revoked';
  approval_decree_no: string | null;
  reviewed_at: string | null;
  vc_credential_id: string | null;
  // Owner display: only first + family. Middle names masked as bullets.
  owner_display_name: string;
  // Polygon as GeoJSON for the verify map; null on legacy deeds without a polygon.
  boundary_polygon_geojson: Record<string, unknown> | null;
  // Public deed PDF link if the deed is configured as publicly downloadable.
  deed_pdf_signed_url: string | null;
}

@Injectable()
export class VerifyService {
  constructor(
    private readonly supabase: SupabaseService,
    private readonly storage: StorageService,
  ) {}

  async byPropertyCode(code: string): Promise<PublicDeedView> {
    const propertyCode = code.trim();
    if (!propertyCode) throw SijilliErrors.notFound('السند العقاري');

    const { data, error } = await this.supabase.admin
      .rpc('property_review_view', { p_property_id: null })
      .maybeSingle();
    // The Phase-5 RPC takes a UUID, not a property_code, so for the
    // public path we do a plain table read by code instead.
    if (error && error.code !== 'PGRST116') {
      throw SijilliErrors.upstream(error.message);
    }
    if (data) {
      // unreachable (we passed null) — kept to silence unused warning
    }

    const property = await this.supabase.admin
      .from('properties')
      .select(
        'id, property_code, parcel_number, property_type, area_sqm, status, ' +
          'approval_decree_no, reviewed_at, vc_credential_id, deed_pdf_path, owner_citizen_id',
      )
      .eq('property_code', propertyCode)
      .eq('status', 'approved')
      .maybeSingle();
    if (property.error) throw SijilliErrors.upstream(property.error.message);
    if (!property.data) throw SijilliErrors.notFound('السند العقاري');
    const p = property.data as unknown as {
      id: string;
      property_code: string;
      parcel_number: string | null;
      property_type: string;
      area_sqm: number | null;
      status: string;
      approval_decree_no: string | null;
      reviewed_at: string | null;
      vc_credential_id: string | null;
      deed_pdf_path: string | null;
      owner_citizen_id: string;
    };

    // Owner display: first + family only.
    const owner = await this.supabase.admin
      .from('citizens')
      .select('first_name_ar, father_name_ar, grandfather_name_ar, family_name_ar')
      .eq('id', p.owner_citizen_id)
      .maybeSingle();
    if (owner.error) throw SijilliErrors.upstream(owner.error.message);
    const o = (owner.data ?? {}) as {
      first_name_ar?: string;
      father_name_ar?: string | null;
      grandfather_name_ar?: string | null;
      family_name_ar?: string;
    };
    const ownerDisplay = [
      o.first_name_ar ?? '—',
      maskName(o.father_name_ar),
      maskName(o.grandfather_name_ar),
      o.family_name_ar ?? '',
    ]
      .filter(Boolean)
      .join(' ');

    // Polygon for the public map. We use a separate RPC so we don't
    // expose internal columns from properties.
    const polygon = await this.supabase.admin
      .rpc('property_polygon_geojson', { p_property_id: p.id })
      .maybeSingle();
    const polygonGeoJson =
      polygon.data && typeof polygon.data === 'object'
        ? (polygon.data as Record<string, unknown>)
        : null;

    // Public deed download path. The verify controller exposes a streaming
    // GET /verify/:code/deed.pdf endpoint that proxies through StorageService;
    // signed URLs (Supabase) are no longer in play under local SQL Server.
    const deedSignedUrl = p.deed_pdf_path ? `/api/v1/verify/${p.property_code}/deed.pdf` : null;

    return {
      property_code: p.property_code,
      parcel_number: p.parcel_number,
      property_type: p.property_type,
      area_sqm: p.area_sqm,
      status: 'active',
      approval_decree_no: p.approval_decree_no,
      reviewed_at: p.reviewed_at,
      vc_credential_id: p.vc_credential_id,
      owner_display_name: ownerDisplay,
      boundary_polygon_geojson: polygonGeoJson,
      deed_pdf_signed_url: deedSignedUrl,
    };
  }
}

// Replace each character of a middle name with a bullet so the deed
// cannot be used to recover full names. Empty strings stay empty.
function maskName(s: string | null | undefined): string {
  if (!s || s.length === 0) return '';
  // Keep first character, mask the rest with bullets.
  return s.charAt(0) + '•'.repeat(Math.max(0, s.length - 1));
}

File: apps/api-dotnet/Verify/VerifyService.cs
using System.Data;
using System.Text.Json;
using Microsoft.Data.SqlClient;
using Microsoft.EntityFrameworkCore;
using Sarh.Api.Blockchain;
using Sarh.Api.Common.Errors;
using Sarh.Api.Data;

namespace Sarh.Api.Verify;

// Public verification of a deed by property_code. The response is a
// SANITIZED view: only the citizen's first and family names are returned
// in full; middle names are masked. PII (phone, dob, etc.) never leaves
// the API. Endpoint is unauthenticated by design — verify QRs are public.
public sealed class VerifyService(
    SarhDbContext db,
    IBlockchainService chain,
    IIpfsService ipfs)
{
    // Statuses that count as "publicly verifiable". Adds 'minted' +
    // 'transferred' on top of the legacy 'approved' since both imply
    // the deed has been signed and (additionally) anchored on-chain.
    private static readonly string[] PublicStatuses = ["approved", "minted", "transferred"];

    public async Task<PublicDeedView> ByPropertyCodeAsync(string code, CancellationToken ct)
    {
        var propertyCode = code.Trim();
        if (string.IsNullOrEmpty(propertyCode))
            throw SarhException.NotFound("السند العقاري", "Deed");

        var p = await db.Properties.AsNoTracking()
            .Where(x => x.PropertyCode == propertyCode && PublicStatuses.Contains(x.Status))
            .Select(x => new
            {
                x.Id,
                x.PropertyCode,
                x.ParcelNumber,
                x.PropertyType,
                x.AreaSqm,
                x.Status,
                x.ApprovalDecreeNo,
                x.ReviewedAt,
                x.VcCredentialId,
                x.DeedPdfPath,
                x.DeedSignedHash,
                x.OwnerCitizenId,
            })
            .FirstOrDefaultAsync(ct)
            ?? throw SarhException.NotFound("السند العقاري", "Deed");

        var owner = await db.Citizens.AsNoTracking()
            .Where(c => c.Id == p.OwnerCitizenId)
            .Select(c => new { c.FirstNameAr, c.FatherNameAr, c.GrandfatherNameAr, c.FamilyNameAr })
            .FirstOrDefaultAsync(ct);

        var ownerDisplay = string.Join(" ", new[]
        {
            owner?.FirstNameAr is { Length: > 0 } first ? first : "—",
            MaskName(owner?.FatherNameAr),
            MaskName(owner?.GrandfatherNameAr),
            owner?.FamilyNameAr ?? "",
        }.Where(s => !string.IsNullOrEmpty(s)));

        var polygon = await LoadPolygonGeoJsonAsync(p.Id, ct);

        var deedSignedUrl = !string.IsNullOrEmpty(p.DeedPdfPath)
            ? $"/api/v1/verify/{p.PropertyCode}/deed.pdf"
            : null;

        // On-chain NFT (if any). Look up the active row only — a 'failed' or
        // 'burned' NFT must not appear on the public deed view.
        var nft = await db.PropertyNfts.AsNoTracking()
            .Where(n => n.PropertyId == p.Id && (n.Status == "minted" || n.Status == "transferred" || n.Status == "pending"))
            .OrderByDescending(n => n.MintedAt)
            .FirstOrDefaultAsync(ct);

        PublicNftView? nftView = null;
        if (nft is not null)
        {
            // Read live owner from chain (stub returns null — treated as
            // "not reconciled" downstream, NOT as a mismatch).
            string? onChainOwner = null;
            try { onChainOwner = await chain.OwnerOfAsync(nft.TokenId, ct); }
            catch { /* explorer/RPC outage shouldn't break verify */ }

            bool? matches = onChainOwner is null
                ? null
                : string.Equals(onChainOwner, nft.OwnerAddress, StringComparison.OrdinalIgnoreCase);

            nftView = new PublicNftView
            {
                TokenId = nft.TokenId,
                ContractAddress = nft.ContractAddress,
                Network = nft.Network,
                Standard = nft.Standard,
                OwnerDid = nft.OwnerDid,
                OwnerAddress = nft.OwnerAddress,
                MetadataUri = nft.MetadataUri,
                MetadataGatewayUrl = ipfs.GatewayUrlFor(nft.MetadataUri),
                MintTxHash = nft.MintTxHash,
                ExplorerTxUrl = chain.ExplorerTxUrl(nft.MintTxHash),
                ExplorerTokenUrl = chain.ExplorerTokenUrl(nft.TokenId),
                MintedAt = nft.MintedAt,
                Status = nft.Status,
                OnChainOwnerMatches = matches,
                OnChainOwnerAddress = onChainOwner,
            };
        }

        return new PublicDeedView
        {
            PropertyCode = p.PropertyCode!,
            ParcelNumber = p.ParcelNumber,
            PropertyType = p.PropertyType,
            AreaSqm = p.AreaSqm,
            Status = p.Status,
            ApprovalDecreeNo = p.ApprovalDecreeNo,
            ReviewedAt = p.ReviewedAt,
            VcCredentialId = p.VcCredentialId,
            OwnerDisplayName = ownerDisplay,
            BoundaryPolygonGeojson = polygon,
            DeedPdfSignedUrl = deedSignedUrl,
            DeedSignedHash = p.DeedSignedHash,
            Nft = nftView,
        };
    }

    public async Task<(string PropertyCode, string DeedPdfPath, string? DeedSignedHash)> ResolveDeedPathAsync(string code, CancellationToken ct)
    {
        var propertyCode = code.Trim();
        var row = await db.Properties.AsNoTracking()
            .Where(p => p.PropertyCode == propertyCode && PublicStatuses.Contains(p.Status))
            .Select(p => new { p.PropertyCode, p.DeedPdfPath, p.DeedSignedHash })
            .FirstOrDefaultAsync(ct);
        if (row is null || string.IsNullOrEmpty(row.DeedPdfPath))
            throw SarhException.NotFound("السند العقاري", "Deed");
        return (row.PropertyCode!, row.DeedPdfPath!, row.DeedSignedHash);
    }

    private async Task<JsonElement?> LoadPolygonGeoJsonAsync(Guid propertyId, CancellationToken ct)
    {
        var conn = (SqlConnection)db.Database.GetDbConnection();
        if (conn.State != ConnectionState.Open) await conn.OpenAsync(ct);
        await using var cmd = conn.CreateCommand();
        cmd.CommandText = "EXEC dbo.property_polygon_geojson @p_property_id;";
        cmd.Parameters.Add(new SqlParameter("@p_property_id", SqlDbType.UniqueIdentifier) { Value = propertyId });

        var raw = await cmd.ExecuteScalarAsync(ct);
        if (raw is null or DBNull) return null;
        var s = raw.ToString();
        if (string.IsNullOrEmpty(s)) return null;
        try
        {
            using var doc = JsonDocument.Parse(s);
            return doc.RootElement.Clone();
        }
        catch (JsonException) { return null; }
    }

    // Replace each character of a middle name (after the first) with a
    // bullet so the deed cannot be used to recover the full name.
    private static string MaskName(string? s)
    {
        if (string.IsNullOrEmpty(s)) return "";
        return s[0] + new string('•', Math.Max(0, s.Length - 1));
    }
}
